service/history/tasks/range.go

```diff
@@ -0,0 +1,114 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"fmt"
+)
+
+type (
+	Range struct {
+		InclusiveMin Key
+		ExclusiveMax Key
+	}
+)
+
+func NewRange(
+	inclusiveMin Key,
+	exclusiveMax Key,
+) Range {
+	if inclusiveMin.CompareTo(exclusiveMax) > 0 {
+		panic(fmt.Sprintf("invalid task range, min %v is larger than max %v", inclusiveMin, exclusiveMax))
+	}
+
+	return Range{
+		InclusiveMin: inclusiveMin,
+		ExclusiveMax: exclusiveMax,
+	}
+}
+
+func (r *Range) IsEmpty() bool {
+	return r.InclusiveMin.CompareTo(r.ExclusiveMax) == 0
+}
+
+func (r *Range) ContainsKey(
+	key Key,
+) bool {
+	return key.CompareTo(r.InclusiveMin) >= 0 &&
+		key.CompareTo(r.ExclusiveMax) < 0
+}
```

Comment:
if InclusiveMin == ExclusiveMax == key, should this function return true or false? <- interesting Q

Reply:
i believe this should return false, which is the current impl

Reply:
Yeah I agree. An empty range should contain no key.